File: api/Ganymede.Api.Data/Monsters/OptionalStats/Languages/MonsterLanguage.cs
using Ganymede.Api.Data.Common;
using System.ComponentModel.DataAnnotations.Schema;

namespace Ganymede.Api.Data.Monsters.OptionalStats.Languages
{
    public class MonsterLanguage
    {
        public bool Understand { get; set; }
        public bool Speak { get; set; }
        public bool Read { get; set; }
        public bool Write { get; set; }

        public int LanguageID { get; set; }
        [ForeignKey("LanguageID")]
        public Language Language { get; set; }

        public int MonsterLanguageSetID { get; set; }
        [ForeignKey("MonsterLanguageSetID")]
        public MonsterLanguageSet MonsterLanguageSet { get; set; }
    }
}

File: src/app/campaign/modules/monster/models/optionalStats/languages/monsterLanguage.ts
import { Language } from "src/app/models/core/language";

export class MonsterLanguage {
	public language: Language;
	public understand: boolean;
	public speak: boolean;
	public read: boolean;
	public write: boolean;

	public static getDefault(): MonsterLanguage {
		const ml = new MonsterLanguage();

		ml.language = Language.getDefault();

		return ml;
	}

	public static isEqual(a: MonsterLanguage, b: MonsterLanguage): boolean {
		if ((!a && b) || (a && !b)) return false;
		if (!a && !b) return true;

		return (
			a === b ||
			(Language.isEqual(a.language, b.language) &&
				a.understand === b.understand &&
				a.speak === b.speak &&
				a.read === b.read &&
				a.write === b.write)
		);
	}

	public static areEqual(
		a: MonsterLanguage[],
		b: MonsterLanguage[]
	): boolean {
		let areLanguagesSame: boolean = true;
		if (a && b) {
			if (a.length === b.length)
				for (let i = 0; i < a.length; i++) {
					if (i > b.length) {
						areLanguagesSame = false;
						break;
					}

					if (!this.isEqual(a[i], b[i])) {
						areLanguagesSame = false;
						break;
					}
				}
			else areLanguagesSame = false;
		} else areLanguagesSame = false;

		return areLanguagesSame;
	}
}
